### PR TITLE
bpo-26256: Document algorithm speed for the Decimal module.

### PR DESCRIPTION
**Pull request by csabella**
https://bugs.python.org/issue26256

### CODE REVIEW

File: Doc/library/decimal.rst [hunk 1]
```diff
@@ -2115,3 +2115,17 @@ Alternatively, inputs can be rounded upon creation using the
 
    >>> Context(prec=5, rounding=ROUND_DOWN).create_decimal('1.2345678')
    Decimal('1.2345')
+
+Q. Is the CPython implementation fast for large numbers?
+
+A. Yes.  A change was added to the Cpython specific implementation that uses
```
Comment (vstinner) [inline Doc/library/decimal.rst]: Cpython => CPython

**Review by vstinner — APPROVED** [review]
LGTM. But I would prefer that @skrah review it, since I'm not use of the context.

@skrah: Would you mind to review this change?

File: Doc/library/decimal.rst [hunk 2]
```diff
@@ -2115,3 +2115,17 @@ Alternatively, inputs can be rounded upon creation using the
 
    >>> Context(prec=5, rounding=ROUND_DOWN).create_decimal('1.2345678')
    Decimal('1.2345')
+
```
Comment (skrah) [inline Doc/library/decimal.rst:2118]: The speedups are in `libmpdec`. Recent versions of `pypy3` also use `libmpdec`, so they have the feature, too (I checked `PyPy 5.8.0-beta0`).

So I'd mention `libmpdec`, `CPython` and `PyPy3`.

In order to give people a more accurate idea of the algorithms used, I'd mention that `libmpdec` uses `Karatsuba multiplication` for medium sized numbers and the `Number Theoretic Transform` for very large numbers.

Comment (csabella) [reply Doc/library/decimal.rst:2118]: Thanks @skrah.  I've included your recommendations in the latest commit.

File: Doc/library/decimal.rst [hunk 3]
```diff
@@ -2115,3 +2115,23 @@ Alternatively, inputs can be rounded upon creation using the
...
+Q. Is the CPython implementation fast for large numbers?
+
+A. Yes.  In the CPython and PyPy3 implementations, the C version of
```
Comment (skrah) [inline Doc/library/decimal.rst]: Thanks, that looks great!  Small nitpick:  C/CFFI versions.  PyPy uses CFFI, I think that is a little clearer.

Comment (csabella) [reply Doc/library/decimal.rst]: Thank Stefan!  I've added that change.

**Comment by skrah** [comment]
Ok, thanks!

**Comment by vstinner** [comment]
I'm sorry but I don't understand the added FAQ entry. Does it mean that decimal is slow by default unless I enable **unlimited** precision?

```
import decimal
import tracemalloc
tracemalloc.start()
if 0:
    ctx = decimal.getcontext()
    ctx.prec = decimal.MAX_PREC
    ctx.Emax = decimal.MAX_EMAX
    ctx.Emin = decimal.MIN_EMIN
num = decimal.Decimal(1) / decimal.Decimal(7)
mem = tracemalloc.get_traced_memory()[1]
print("Peak memory usage: %.1f kB" % mem)
```

On my Fedora 29, this script displays "Peak memory usage: 864.0 kB".

But if I replace "if 0:" with "if 1:", the script fails immediately with:

```
Traceback (most recent call last):
  File "dec.py", line 9, in <module>
    num = decimal.Decimal(1) / decimal.Decimal(7)
MemoryError
```

**Comment by skrah** [comment]
On Tue, Feb 12, 2019 at 06:17:15AM -0800, Victor Stinner wrote:
> Does it mean that decimal is slow by default unless I enable **unlimited** precision?

No, but I can see why you get the impression from the new FAQ entry. I hoped
people would infer the intended meaning.

"to realize this performance gain" here means: If you don't set for
unrounded precision, you'll simply get a result that's rounded to,
say, 28 digits instead of the desired unrounded integer.


It could be replaced by "in order to get an unrounded exact integer result".


> But if I replace "if 0:" with "if 1:", the script fails immediately with:
> 
> ```
> Traceback (most recent call last):
>   File "dec.py", line 9, in <module>
>     num = decimal.Decimal(1) / decimal.Decimal(7)
> MemoryError

Yes, because you don't have enough memory for decimal.MAX_PREC digits.

**Comment by vstinner** [comment]
> However, to realize this performance gain, the context needs to be set for unrounded calculations.

I understand: if you don't use "ctx.prec = decimal.MAX_PREC", decimal will be very slow.

> It could be replaced by "in order to get an unrounded exact integer result".

I'm confused. The FAQ entry is about performance. I don't understand the relationship between performance and correctness.

**Comment by skrah** [comment]
On Tue, Feb 12, 2019 at 06:48:57AM -0800, Victor Stinner wrote:
> > However, to realize this performance gain, the context needs to be set for unrounded calculations.
> 
> I understand: if you don't use "ctx.prec = decimal.MAX_PREC", decimal will be very slow.
> 
> > It could be replaced by "in order to get an unrounded exact integer result".
> 
> I'm confused. The FAQ entry is about performance. I don't understand the relationship between performance and correctness.

Then why did you approve the original entry? This is just wasting time.

**Comment by vstinner** [comment]
> Then why did you approve the original entry?

Sorry, I don't recall the history of this PR created longer than 1 year ago. It seems like the PR evolved after I approved it, and I asked you explicitly to review you since I don't know well decimal internals.

I approved the idea of documenting decimal performance, as discussed in https://bugs.python.org/issue26256

> This is just wasting time.

I don't think that it's useful to be rude here.

**Comment by vstinner** [comment]
>  "to realize this performance gain" here means: If you don't set for unrounded precision, you'll simply get a result that's rounded to, say, 28 digits instead of the desired unrounded integer.

Maybe the FAQ entry can be rephrased to avoid confusion. But since I still don't understand very well the subtle details, I cannot suggest a change.

**Pull request merged**